tutorials/haskell/servant-auth/tutorial.md

```diff
@@ -0,0 +1,566 @@
+[Servant web-framework](https://haskell-servant.github.io/) is expected.
+
+## Setting the goal
+
+Currently
+[the Servant docs](http://haskell-servant.readthedocs.io/en/stable/tutorial/Authentication.html)
+describe two ways to perform authentication with the framework: using
+[basic authentication](https://en.wikipedia.org/wiki/Basic_access_authentication)
+and with the above-mentioned generalized authentication. The both methods
+are similar in how you use them in your code, yet generalized authentication
+opens the possibility to read the cookie you might have previously set,
+because it has access to request data.
+
+In this tutorial we will set and then use for authentication an encrypted
+cookie with (possibly) arbitrary session data in it. All the session data
+will be stored client-side. It seems to make sense to check cookie and
+return a `Maybe Session` value to give the handlers more freedom in how to
+use session data or the fact of its absence (for example, we will change the
+application menu depending on whether the user is logged in or not — a
+reasonable thing to expect from a web application).
+
+## Imports and language extensions
+
+Let's start by importing some modules and enabling a few language
+extensions. The complete source code of this application is available in the
+[Stack Builders's tutorials repository on GitHub](https://github.com/stackbuilders/tutorials),
+so you can just clone the repository and start playing with it.
+
+```haskell
+{-# LANGUAGE DataKinds            #-}
+{-# LANGUAGE FlexibleInstances    #-}
+{-# LANGUAGE OverloadedStrings    #-}
+{-# LANGUAGE RecordWildCards      #-}
+{-# LANGUAGE ScopedTypeVariables  #-}
+{-# LANGUAGE TypeFamilies         #-}
+{-# LANGUAGE TypeOperators        #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Main (main) where
+
+import Control.Monad.Catch (try)
+import Control.Monad.Except
+import Control.Monad.Reader
+import Crypto.Cipher.AES (AES256)
+import Crypto.Cipher.Types (ctrCombine)
+import Crypto.Hash.Algorithms (SHA256)
+import Crypto.Random
+import Data.ByteString (ByteString)
+import Data.Serialize hiding (Get)
+import Data.Text (Text)
+import Lucid
+import Network.URI hiding (scheme)
+import Network.Wai
+import Network.Wai.Handler.Warp
+import Servant
+import Servant.HTML.Lucid
+import Servant.Server.Experimental.Auth
+import Servant.Server.Experimental.Auth.Cookie
+import qualified Data.ByteString    as B
+import qualified Data.Text          as T
+import qualified Data.Text.Encoding as T
+```
+
+## The handler monad
+
+To work with cookies, we will need to be able to access at least three
+values from handlers: `AuthCookieSettings` (more on that below),
+`RandomSource`, and `ServerKey`. For type-safe rendering we will need three
+more things: approot (as a `String`), port (`Int`), and scheme (`String`).
+Knowing these requirements it's an obvious choice to use monad stack with
+`ReaderT` parametrized over `AppContext` record containing all necessary
```

Comment:
@mrkkrp "to use a monad stack with `ReaderT` parametrized over the `AppContext` record"